{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "68a8de9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import bmt\n",
    "import requests\n",
    "import linkml\n",
    "from linkml_runtime import SchemaView\n",
    "from collections import defaultdict\n",
    "from typing import Tuple, List\n",
    "import requests\n",
    "import yaml\n",
    "from pprint import pprint\n",
    "from biothings_explorer.smartapi_kg.dataload import load_specs\n",
    "import pandas\n",
    "\n",
    "tk = bmt.Toolkit('https://raw.githubusercontent.com/biolink/biolink-model/2.2.12/biolink-model.yaml')\n",
    "tsv_file = open(\"predicates.tsv\", \"w\")\n",
    "tsv_attributes = open(\"attributes.tsv\", \"w\")\n",
    "tsv_writer = csv.writer(tsv_file, delimiter='\\t')\n",
    "tsv_writer_att = csv.writer(tsv_attributes, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e56dc0d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def sample_predicates():\n",
    "    specs = load_specs()\n",
    "    kp_titles = []\n",
    "    for spec in specs:\n",
    "        if 'x-translator' not in spec['info']:\n",
    "            continue\n",
    "        if spec['info']['x-translator']['component'] == 'KP':\n",
    "            kp_titles.append(spec['info']['title'])\n",
    "        if 'servers' not in spec:\n",
    "            continue\n",
    "        else:\n",
    "            url = spec['servers'][0]['url']\n",
    "            apititle = '_'.join(spec['info']['title'].split())\n",
    "            if url.endswith('/'):\n",
    "                url = url[:-1]\n",
    "            predicates_url = f'{url}/meta_knowledge_graph'\n",
    "            print(predicates_url)\n",
    "            trapi, predicates = get_predicates(predicates_url)\n",
    "            # if is_trapi:\n",
    "            if not predicates:\n",
    "                continue\n",
    "            else:\n",
    "                print(\"got predicates\")\n",
    "                preds, attribs, url = dump_trapi_predicate_results(predicates, predicates_url)\n",
    "                predicates = set(preds)\n",
    "                attributes = set(attribs)\n",
    "                for pred in predicates:\n",
    "                    tsv_writer.writerow([apititle, url, pred])\n",
    "                for attrib in attributes:\n",
    "                    tsv_writer_att.writerow([apititle, url, attrib])\n",
    "            # else:\n",
    "            #     dump_smartapi_predicate_results(spec['info']['title'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "01040319",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def in_biolink_model(predicate):\n",
    "    is_predicate = tk.is_predicate(predicate)\n",
    "    return is_predicate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "09086adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def dump_trapi_predicate_results(predicates, url):\n",
    "    preds = []\n",
    "    attribs = []\n",
    "    for edge in predicates.get('edges'):\n",
    "        predicate = edge.get('predicate')\n",
    "        subject = edge.get('subject')\n",
    "        tobject = edge.get('object')\n",
    "        if 'attributes' in edge and edge.get('attributes') is not None:\n",
    "            for attribute in edge.get('attributes'):\n",
    "                attribs.append(attribute.get('attribute_type_id'))\n",
    "        preds.append(predicate)\n",
    "    return preds, attribs, url\n",
    "\n",
    "            # tsv_writer.writerow([subject, predicate, tobject, url])\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "86b86b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_predicates(pr_url):\n",
    "    try:\n",
    "        response = requests.get(pr_url)\n",
    "        if response.status_code == 200:\n",
    "            return True, response.json()\n",
    "        else:\n",
    "            return False, {}\n",
    "    except:\n",
    "        return False, {}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "383c74f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/socio_environmental_exposures_api/v1/meta_knowledge_graph\n",
      "/cmaq_exposures_api/v1/meta_knowledge_graph\n",
      "/roadway_proximity_api/v1/meta_knowledge_graph\n",
      "https://name-resolution-sri.renci.org/meta_knowledge_graph\n",
      "https://icees-pcd.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://icees-dili-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/proteins/api/meta_knowledge_graph\n",
      "https://icees-asthma-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://icees-asthma.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://translator-workflow-runner.renci.org/meta_knowledge_graph\n",
      "https://icees-pcd-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/ols/api/meta_knowledge_graph\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1/meta_knowledge_graph\n",
      "https://icees-dili.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/QuickGO/services/meta_knowledge_graph\n",
      "https://ontology-kp.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://trapi.monarchinitiative.org/1.2/meta_knowledge_graph\n",
      "https://medikanren-trapi.ci.transltr.io/v2/meta_knowledge_graph\n",
      "https://automat.renci.org/gtex/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/intact/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/gtopdb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/drugcentral/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/chemical-normalization/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/ctd/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/cord19/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/panther/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/textminingkp/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://bl-lookup-sri.renci.org/meta_knowledge_graph\n",
      "https://automat.renci.org/hmdb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/covidkopkg/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://nodenormalization-sri.renci.org/1.2/meta_knowledge_graph\n",
      "https://nodenormalization-dev.apps.renci.org/1.2/meta_knowledge_graph\n",
      "https://automat.renci.org/pharos/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/human-goa/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/uberongraph/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/biolink/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/viral-proteome/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cam-kp-api.renci.org/1.2.0/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/hgnc/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cam-kp-api.renci.org/1.2.0/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/robokopkg/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/ontological-hierarchy/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/foodb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/gwas-catalog/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/molepro-fda/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/hetio/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://explanatory-agent.azurewebsites.net/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/uberon/meta_knowledge_graph\n",
      "https://aragorn-dev.apps.renci.org/1.2/meta_knowledge_graph\n",
      "http://chp.thayer.dartmouth.edu/meta_knowledge_graph\n",
      "got predicates\n",
      "https://smart-api.info/api/meta_knowledge_graph\n",
      "https://biothings.ncats.io/hpo/meta_knowledge_graph\n",
      "https://aragorn.renci.org/1.2/meta_knowledge_graph\n",
      "https://cooccurrence-p36smkc6hq-uc.a.run.app/meta_knowledge_graph\n",
      "got predicates\n",
      "https://spokekp.healthdatascience.cloud/api/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cohd.io/api/meta_knowledge_graph\n",
      "got predicates\n",
      "https://openpredict.semanticscience.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://mygene.info/v3/meta_knowledge_graph\n",
      "https://myvariant.info/v1/meta_knowledge_graph\n",
      "https://biothings.ncats.io/rhea/meta_knowledge_graph\n",
      "https://molepro-trapi.transltr.io/molepro/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://ia.healthdatascience.cloud/api/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "http://mydisease.info/v1/meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/DISEASES/meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/mgigene2phenotype/meta_knowledge_graph\n",
      "https://mychem.info/v1/meta_knowledge_graph\n",
      "https://biothings.ncats.io/ebigene2phenotype/meta_knowledge_graph\n",
      "https://biothings.ncats.io/go_bp/meta_knowledge_graph\n",
      "https://arax.ncats.io/beta/api/rtxkg2/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/go_cc/meta_knowledge_graph\n",
      "https://arax.ncats.io/api/rtxkg2/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://api.bte.ncats.io/v1/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/idisk/meta_knowledge_graph\n",
      "https://biothings.ncats.io/go_mf/meta_knowledge_graph\n",
      "https://api.bte.ncats.io/v1/team/Service%20Provider/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/multiomics_wellness_kp/meta_knowledge_graph\n",
      "https://api.collaboratory.semanticscience.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/dgidb/meta_knowledge_graph\n",
      "https://biothings.ncats.io/semmeddb/meta_knowledge_graph\n",
      "https://biothings.ncats.io/drug_response_kp/meta_knowledge_graph\n",
      "https://api.monarchinitiative.org/api/meta_knowledge_graph\n",
      "https://biothings.ncats.io/text_mining_targeted_association/meta_knowledge_graph\n",
      "https://arax.ncats.io/api/arax/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/clinical_risk_kp/meta_knowledge_graph\n",
      "https://arax.ncats.io/beta/api/arax/v1.2/meta_knowledge_graph\n",
      "got predicates\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if __name__ == '__main__':\n",
    "    sample_predicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d26f6993",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv = SchemaView('https://raw.githubusercontent.com/biolink/biolink-model/master/biolink-model.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "368fa0b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "related to\n",
      "['related to at concept level', 'related to at instance level', 'disease has location', 'composed primarily of']\n",
      "related to at concept level\n",
      "['superclass of', 'subclass of', 'close match', 'broad match', 'narrow match']\n",
      "related to at instance level\n",
      "['associated with', 'opposite of', 'has real world evidence of association with', 'target for', 'has target', 'active in', 'acts upstream of', 'has upstream actor', 'mentions', 'mentioned by', 'contributor', 'has contributor', 'interacts with', 'affects', 'affected by', 'diagnoses', 'is diagnosed by', 'increases amount or activity of', 'decreases amount or activity of', 'gene product of', 'has gene product', 'transcribed to', 'transcribed from', 'translates to', 'translation of', 'coexists with', 'genetic association', 'affects risk for', 'risk affected by', 'contributes to', 'contribution from', 'caused by', 'treated by', 'has phenotype', 'phenotype of', 'occurs in', 'contains process', 'located in', 'location of', 'similar to', 'has sequence location', 'sequence location of', 'model of', 'models', 'overlaps', 'has participant', 'participates in', 'derives into', 'derives from', 'manifestation of', 'has manifestation', 'produces', 'produced by', 'temporally related to', 'related condition', 'is sequence variant of', 'has sequence variant', 'disease has basis in', 'occurs in disease', 'contraindicated for', 'has contraindication', 'has not completed', 'not completed by', 'has completed', 'completed by', 'in linkage disequilibrium with', 'has increased amount', 'increased amount of', 'has decreased amount', 'decreased amount in', 'lacks part', 'missing from', 'develops from', 'develops into', 'in taxon', 'taxon of', 'has molecular consequence', 'is molecular consequence of']\n",
      "associated with\n",
      "['associated with sensitivity to', 'sensitivity associated with', 'associated with resistance to', 'resistance associated with', 'correlated with']\n",
      "superclass of\n",
      "[]\n",
      "subclass of\n",
      "[]\n",
      "same as\n",
      "[]\n",
      "close match\n",
      "['exact match']\n",
      "exact match\n",
      "['same as']\n",
      "broad match\n",
      "[]\n",
      "narrow match\n",
      "[]\n",
      "opposite of\n",
      "[]\n",
      "has real world evidence of association with\n",
      "[]\n",
      "target for\n",
      "[]\n",
      "has target\n",
      "[]\n",
      "active in\n",
      "[]\n",
      "acts upstream of\n",
      "['acts upstream of positive effect', 'acts upstream of negative effect', 'acts upstream of or within', 'acts upstream of or within positive effect', 'acts upstream of or within negative effect']\n",
      "has upstream actor\n",
      "['has positive upstream actor', 'has negative upstream actor', 'has upstream or within actor']\n",
      "acts upstream of positive effect\n",
      "[]\n",
      "has positive upstream actor\n",
      "[]\n",
      "acts upstream of negative effect\n",
      "[]\n",
      "has negative upstream actor\n",
      "[]\n",
      "acts upstream of or within\n",
      "[]\n",
      "has upstream or within actor\n",
      "['has positive upstream or within actor', 'has negative upstream or within actor']\n",
      "acts upstream of or within positive effect\n",
      "[]\n",
      "has positive upstream or within actor\n",
      "[]\n",
      "acts upstream of or within negative effect\n",
      "[]\n",
      "has negative upstream or within actor\n",
      "[]\n",
      "mentions\n",
      "[]\n",
      "mentioned by\n",
      "[]\n",
      "contributor\n",
      "['provider', 'publisher', 'editor', 'author']\n",
      "has contributor\n",
      "['has provider', 'has publisher', 'has editor', 'has author']\n",
      "provider\n",
      "[]\n",
      "has provider\n",
      "[]\n",
      "publisher\n",
      "[]\n",
      "has publisher\n",
      "[]\n",
      "editor\n",
      "[]\n",
      "has editor\n",
      "[]\n",
      "author\n",
      "[]\n",
      "has author\n",
      "[]\n",
      "interacts with\n",
      "['physically interacts with', 'genetically interacts with', 'directly interacts with']\n",
      "physically interacts with\n",
      "['chemically interacts with', 'molecularly interacts with']\n",
      "chemically interacts with\n",
      "[]\n",
      "molecularly interacts with\n",
      "['decreases molecular interaction', 'molecular interaction decreased by', 'increases molecular interaction', 'molecular interaction increased by']\n",
      "genetically interacts with\n",
      "[]\n",
      "affects\n",
      "['affects abundance of', 'affects activity of', 'affects expression of', 'affects folding of', 'affects localization of', 'affects metabolic processing of', 'affects molecular modification of', 'affects synthesis of', 'affects degradation of', 'affects mutation rate of', 'affects response to', 'affects splicing of', 'affects stability of', 'affects transport of', 'affects secretion of', 'affects uptake of', 'process regulates process', 'entity regulates entity', 'disrupts', 'ameliorates', 'exacerbates', 'affects expression in']\n",
      "affected by\n",
      "['abundance affected by', 'activity affected by', 'expression affected by', 'folding affected by', 'localization affected by', 'metabolic processing affected by', 'molecular modification affected by', 'synthesis affected by', 'degradation affected by', 'mutation rate affected by', 'response affected by', 'splicing affected by', 'stability affected by', 'transport affected by', 'secretion affected by', 'uptake affected by', 'process regulated by process', 'entity regulated by entity', 'disrupted by', 'is ameliorated by', 'is exacerbated by']\n",
      "associated with sensitivity to\n",
      "[]\n",
      "sensitivity associated with\n",
      "[]\n",
      "associated with resistance to\n",
      "[]\n",
      "resistance associated with\n",
      "[]\n",
      "diagnoses\n",
      "[]\n",
      "is diagnosed by\n",
      "[]\n",
      "increases amount or activity of\n",
      "['increases abundance of', 'increases activity of', 'increases expression of', 'increases folding of', 'increases localization of', 'increases metabolic processing of', 'increases molecular modification of', 'increases synthesis of', 'decreases degradation of', 'increases mutation rate of', 'increases response to', 'increases splicing of', 'increases stability of', 'increases transport of', 'increases secretion of', 'increases uptake of', 'increases molecular interaction']\n",
      "decreases amount or activity of\n",
      "['decreases abundance of', 'decreases activity of', 'decreases expression of', 'decreases folding of', 'decreases localization of', 'decreases metabolic processing of', 'decreases molecular modification of', 'decreases synthesis of', 'increases degradation of', 'decreases mutation rate of', 'decreases response to', 'decreases splicing of', 'decreases stability of', 'decreases transport of', 'decreases secretion of', 'decreases uptake of', 'decreases molecular interaction']\n",
      "affects abundance of\n",
      "['increases abundance of', 'decreases abundance of']\n",
      "abundance affected by\n",
      "['abundance increased by', 'abundance decreased by']\n",
      "increases abundance of\n",
      "[]\n",
      "abundance increased by\n",
      "[]\n",
      "decreases abundance of\n",
      "[]\n",
      "abundance decreased by\n",
      "[]\n",
      "increases activity of\n",
      "[]\n",
      "affects activity of\n",
      "['increases activity of', 'decreases activity of']\n",
      "activity affected by\n",
      "['activity increased by', 'activity decreased by']\n",
      "activity increased by\n",
      "[]\n",
      "decreases activity of\n",
      "[]\n",
      "activity decreased by\n",
      "[]\n",
      "affects expression of\n",
      "['increases expression of', 'decreases expression of']\n",
      "expression affected by\n",
      "['expression increased by', 'expression decreased by']\n",
      "increases expression of\n",
      "[]\n",
      "expression increased by\n",
      "[]\n",
      "decreases expression of\n",
      "[]\n",
      "expression decreased by\n",
      "[]\n",
      "affects folding of\n",
      "['increases folding of', 'decreases folding of']\n",
      "folding affected by\n",
      "['folding increased by', 'folding decreased by']\n",
      "increases folding of\n",
      "[]\n",
      "folding increased by\n",
      "[]\n",
      "decreases folding of\n",
      "[]\n",
      "folding decreased by\n",
      "[]\n",
      "affects localization of\n",
      "['increases localization of', 'decreases localization of']\n",
      "localization affected by\n",
      "['localization increased by', 'localization decreased by']\n",
      "increases localization of\n",
      "[]\n",
      "localization increased by\n",
      "[]\n",
      "decreases localization of\n",
      "[]\n",
      "localization decreased by\n",
      "[]\n",
      "affects metabolic processing of\n",
      "['increases metabolic processing of', 'decreases metabolic processing of']\n",
      "metabolic processing affected by\n",
      "['metabolic processing increased by', 'metabolic processing decreased by']\n",
      "increases metabolic processing of\n",
      "[]\n",
      "metabolic processing increased by\n",
      "[]\n",
      "decreases metabolic processing of\n",
      "[]\n",
      "metabolic processing decreased by\n",
      "[]\n",
      "affects molecular modification of\n",
      "['increases molecular modification of', 'decreases molecular modification of']\n",
      "molecular modification affected by\n",
      "['molecular modification increased by', 'molecular modification decreased by']\n",
      "increases molecular modification of\n",
      "[]\n",
      "molecular modification increased by\n",
      "[]\n",
      "decreases molecular modification of\n",
      "[]\n",
      "molecular modification decreased by\n",
      "[]\n",
      "affects synthesis of\n",
      "['increases synthesis of', 'decreases synthesis of']\n",
      "synthesis affected by\n",
      "['synthesis increased by', 'synthesis decreased by']\n",
      "increases synthesis of\n",
      "[]\n",
      "synthesis increased by\n",
      "[]\n",
      "decreases synthesis of\n",
      "[]\n",
      "synthesis decreased by\n",
      "[]\n",
      "affects degradation of\n",
      "['increases degradation of', 'decreases degradation of']\n",
      "degradation affected by\n",
      "['degradation increased by', 'degradation decreased by']\n",
      "increases degradation of\n",
      "[]\n",
      "degradation increased by\n",
      "[]\n",
      "decreases degradation of\n",
      "[]\n",
      "degradation decreased by\n",
      "[]\n",
      "affects mutation rate of\n",
      "['increases mutation rate of', 'decreases mutation rate of']\n",
      "mutation rate affected by\n",
      "['mutation rate increased by', 'mutation rate decreased by']\n",
      "increases mutation rate of\n",
      "[]\n",
      "mutation rate increased by\n",
      "[]\n",
      "decreases mutation rate of\n",
      "[]\n",
      "mutation rate decreased by\n",
      "[]\n",
      "affects response to\n",
      "['increases response to', 'decreases response to']\n",
      "response affected by\n",
      "['response increased by', 'response decreased by']\n",
      "increases response to\n",
      "[]\n",
      "response increased by\n",
      "[]\n",
      "decreases response to\n",
      "[]\n",
      "response decreased by\n",
      "[]\n",
      "affects splicing of\n",
      "['increases splicing of', 'decreases splicing of']\n",
      "splicing affected by\n",
      "['splicing increased by', 'splicing decreased by']\n",
      "increases splicing of\n",
      "[]\n",
      "splicing increased by\n",
      "[]\n",
      "decreases splicing of\n",
      "[]\n",
      "splicing decreased by\n",
      "[]\n",
      "affects stability of\n",
      "['increases stability of', 'decreases stability of']\n",
      "stability affected by\n",
      "['stability increased by', 'stability decreased by']\n",
      "increases stability of\n",
      "[]\n",
      "stability increased by\n",
      "[]\n",
      "decreases stability of\n",
      "[]\n",
      "stability decreased by\n",
      "[]\n",
      "affects transport of\n",
      "['increases transport of', 'decreases transport of']\n",
      "transport affected by\n",
      "['transport increased by', 'transport decreased by']\n",
      "increases transport of\n",
      "[]\n",
      "transport increased by\n",
      "[]\n",
      "decreases transport of\n",
      "[]\n",
      "transport decreased by\n",
      "[]\n",
      "affects secretion of\n",
      "['increases secretion of', 'decreases secretion of']\n",
      "secretion affected by\n",
      "['secretion increased by', 'secretion decreased by']\n",
      "increases secretion of\n",
      "[]\n",
      "secretion increased by\n",
      "[]\n",
      "decreases secretion of\n",
      "[]\n",
      "secretion decreased by\n",
      "[]\n",
      "affects uptake of\n",
      "['increases uptake of', 'decreases uptake of']\n",
      "uptake affected by\n",
      "['uptake increased by', 'uptake decreased by']\n",
      "increases uptake of\n",
      "[]\n",
      "uptake increased by\n",
      "[]\n",
      "decreases uptake of\n",
      "[]\n",
      "uptake decreased by\n",
      "[]\n",
      "process regulates process\n",
      "['process positively regulates process', 'process negatively regulates process']\n",
      "process regulated by process\n",
      "['process positively regulated by process', 'process negatively regulated by process']\n",
      "process positively regulates process\n",
      "[]\n",
      "process positively regulated by process\n",
      "[]\n",
      "process negatively regulates process\n",
      "[]\n",
      "process negatively regulated by process\n",
      "[]\n",
      "entity regulates entity\n",
      "['entity positively regulates entity', 'entity negatively regulates entity']\n",
      "entity regulated by entity\n",
      "['entity positively regulated by entity', 'entity negatively regulated by entity']\n",
      "entity positively regulates entity\n",
      "[]\n",
      "entity positively regulated by entity\n",
      "[]\n",
      "entity negatively regulates entity\n",
      "[]\n",
      "entity negatively regulated by entity\n",
      "[]\n",
      "disrupts\n",
      "[]\n",
      "disrupted by\n",
      "[]\n",
      "gene product of\n",
      "[]\n",
      "has gene product\n",
      "[]\n",
      "transcribed to\n",
      "[]\n",
      "transcribed from\n",
      "[]\n",
      "translates to\n",
      "[]\n",
      "translation of\n",
      "[]\n",
      "homologous to\n",
      "['paralogous to', 'orthologous to', 'xenologous to']\n",
      "paralogous to\n",
      "[]\n",
      "orthologous to\n",
      "[]\n",
      "xenologous to\n",
      "[]\n",
      "coexists with\n",
      "['in pathway with', 'in complex with', 'in cell population with', 'colocalizes with']\n",
      "in pathway with\n",
      "[]\n",
      "in complex with\n",
      "[]\n",
      "in cell population with\n",
      "[]\n",
      "colocalizes with\n",
      "[]\n",
      "genetic association\n",
      "['gene associated with condition', 'condition associated with gene']\n",
      "gene associated with condition\n",
      "[]\n",
      "condition associated with gene\n",
      "[]\n",
      "affects risk for\n",
      "['predisposes', 'prevents']\n",
      "risk affected by\n",
      "['prevented by']\n",
      "predisposes\n",
      "[]\n",
      "contributes to\n",
      "['causes']\n",
      "contribution from\n",
      "[]\n",
      "causes\n",
      "['causes adverse event', 'adverse event caused by']\n",
      "caused by\n",
      "[]\n",
      "ameliorates\n",
      "['treats']\n",
      "is ameliorated by\n",
      "[]\n",
      "exacerbates\n",
      "[]\n",
      "is exacerbated by\n",
      "[]\n",
      "treats\n",
      "['approved to treat']\n",
      "treated by\n",
      "['approved for treatment by']\n",
      "approved to treat\n",
      "[]\n",
      "approved for treatment by\n",
      "[]\n",
      "prevents\n",
      "[]\n",
      "prevented by\n",
      "[]\n",
      "correlated with\n",
      "['positively correlated with', 'negatively correlated with', 'occurs together in literature with', 'coexpressed with', 'has biomarker', 'biomarker for']\n",
      "positively correlated with\n",
      "[]\n",
      "negatively correlated with\n",
      "[]\n",
      "occurs together in literature with\n",
      "[]\n",
      "coexpressed with\n",
      "[]\n",
      "has biomarker\n",
      "[]\n",
      "biomarker for\n",
      "[]\n",
      "expressed in\n",
      "[]\n",
      "expresses\n",
      "[]\n",
      "has phenotype\n",
      "[]\n",
      "phenotype of\n",
      "[]\n",
      "occurs in\n",
      "[]\n",
      "contains process\n",
      "[]\n",
      "located in\n",
      "['expressed in']\n",
      "location of\n",
      "['expresses']\n",
      "disease has location\n",
      "[]\n",
      "similar to\n",
      "['homologous to', 'chemically similar to']\n",
      "chemically similar to\n",
      "[]\n",
      "has sequence location\n",
      "[]\n",
      "sequence location of\n",
      "[]\n",
      "model of\n",
      "[]\n",
      "models\n",
      "[]\n",
      "overlaps\n",
      "['has part', 'part of']\n",
      "has part\n",
      "['has plasma membrane part', 'has food component', 'has active ingredient', 'has excipient', 'has variant part']\n",
      "has plasma membrane part\n",
      "[]\n",
      "composed primarily of\n",
      "[]\n",
      "plasma membrane part of\n",
      "[]\n",
      "part of\n",
      "['plasma membrane part of', 'food component of', 'is active ingredient of', 'is excipient of', 'variant part of']\n",
      "has input\n",
      "['consumes']\n",
      "is input of\n",
      "['consumed by']\n",
      "has output\n",
      "[]\n",
      "is output of\n",
      "[]\n",
      "has participant\n",
      "['has input', 'has output', 'has catalyst', 'has substrate', 'actively involves', 'enabled by']\n",
      "catalyzes\n",
      "[]\n",
      "has catalyst\n",
      "[]\n",
      "has substrate\n",
      "[]\n",
      "is substrate of\n",
      "[]\n",
      "participates in\n",
      "['is input of', 'is output of', 'catalyzes', 'is substrate of', 'actively involved in', 'enables']\n",
      "actively involved in\n",
      "['capable of']\n",
      "actively involves\n",
      "['has capability']\n",
      "capable of\n",
      "[]\n",
      "has capability\n",
      "[]\n",
      "enables\n",
      "[]\n",
      "enabled by\n",
      "[]\n",
      "derives into\n",
      "['has metabolite']\n",
      "derives from\n",
      "['is metabolite of']\n",
      "is metabolite of\n",
      "[]\n",
      "has metabolite\n",
      "[]\n",
      "food component of\n",
      "['nutrient of']\n",
      "has food component\n",
      "['has nutrient']\n",
      "nutrient of\n",
      "[]\n",
      "has nutrient\n",
      "[]\n",
      "is active ingredient of\n",
      "[]\n",
      "has active ingredient\n",
      "[]\n",
      "is excipient of\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "has excipient\n",
      "[]\n",
      "manifestation of\n",
      "[]\n",
      "has manifestation\n",
      "[]\n",
      "produces\n",
      "[]\n",
      "produced by\n",
      "[]\n",
      "consumes\n",
      "[]\n",
      "consumed by\n",
      "[]\n",
      "temporally related to\n",
      "['precedes', 'preceded by']\n",
      "precedes\n",
      "[]\n",
      "preceded by\n",
      "[]\n",
      "directly interacts with\n",
      "[]\n",
      "affects expression in\n",
      "[]\n",
      "has variant part\n",
      "[]\n",
      "variant part of\n",
      "[]\n",
      "related condition\n",
      "[]\n",
      "is sequence variant of\n",
      "['is missense variant of', 'is synonymous variant of', 'is nonsense variant of', 'is frameshift variant of', 'is splice site variant of', 'is nearby variant of', 'is non coding variant of']\n",
      "has sequence variant\n",
      "['has missense variant', 'has synonymous variant', 'has nonsense variant', 'has frameshift variant', 'has splice site variant', 'has nearby variant', 'has non coding variant']\n",
      "is missense variant of\n",
      "[]\n",
      "has missense variant\n",
      "[]\n",
      "is synonymous variant of\n",
      "[]\n",
      "has synonymous variant\n",
      "[]\n",
      "is nonsense variant of\n",
      "[]\n",
      "has nonsense variant\n",
      "[]\n",
      "is frameshift variant of\n",
      "[]\n",
      "has frameshift variant\n",
      "[]\n",
      "is splice site variant of\n",
      "[]\n",
      "has splice site variant\n",
      "[]\n",
      "is nearby variant of\n",
      "[]\n",
      "has nearby variant\n",
      "[]\n",
      "is non coding variant of\n",
      "[]\n",
      "has non coding variant\n",
      "[]\n",
      "disease has basis in\n",
      "[]\n",
      "occurs in disease\n",
      "[]\n",
      "causes adverse event\n",
      "[]\n",
      "adverse event caused by\n",
      "[]\n",
      "contraindicated for\n",
      "[]\n",
      "has contraindication\n",
      "[]\n",
      "has not completed\n",
      "[]\n",
      "not completed by\n",
      "[]\n",
      "has completed\n",
      "[]\n",
      "completed by\n",
      "[]\n",
      "decreases molecular interaction\n",
      "[]\n",
      "molecular interaction decreased by\n",
      "[]\n",
      "increases molecular interaction\n",
      "[]\n",
      "molecular interaction increased by\n",
      "[]\n",
      "in linkage disequilibrium with\n",
      "[]\n",
      "has increased amount\n",
      "[]\n",
      "increased amount of\n",
      "[]\n",
      "has decreased amount\n",
      "[]\n",
      "decreased amount in\n",
      "[]\n",
      "lacks part\n",
      "[]\n",
      "missing from\n",
      "[]\n",
      "develops from\n",
      "[]\n",
      "develops into\n",
      "[]\n",
      "in taxon\n",
      "[]\n",
      "taxon of\n",
      "[]\n",
      "has molecular consequence\n",
      "[]\n",
      "is molecular consequence of\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "for s in sv.all_slots().values():\n",
    "    if 'related to' in sv.slot_ancestors(s.name):\n",
    "        print(s.name)\n",
    "        print(sv.slot_children(s.name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5d1e750d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_predicate_tree_data(biolink_version: str) -> Tuple[List[dict], str]:\n",
    "    # Grab Biolink yaml file and load into dictionary tree structures\n",
    "    response = requests.get(f\"https://raw.githubusercontent.com/biolink/biolink-model/\"\n",
    "                            f\"{biolink_version if biolink_version else 'master'}/biolink-model.yaml\",\n",
    "                            timeout=10)\n",
    "    if response.status_code == 200:\n",
    "        # Build predicates tree\n",
    "        biolink_model = yaml.safe_load(response.text)\n",
    "        parent_to_child_dict = defaultdict(set)\n",
    "        for slot_name_english, info in biolink_model[\"slots\"].items():\n",
    "            slot_name = convert_predicate_to_trapi_format(slot_name_english)\n",
    "            parent_name_english = info.get(\"is_a\")\n",
    "            if parent_name_english:\n",
    "                parent_name = convert_predicate_to_trapi_format(parent_name_english)\n",
    "                parent_to_child_dict[parent_name].add(slot_name)\n",
    "        root_node = {\"name\": \"related_to\"}\n",
    "        predicate_tree = get_tree_node_recursive(root_node, parent_to_child_dict)\n",
    "\n",
    "        biolink_version = biolink_model[\"version\"]\n",
    "        return [predicate_tree], biolink_version\n",
    "    else:\n",
    "        return [], \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9f8e5832",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tree_node_recursive(root_node: dict, parent_to_child_map: dict):\n",
    "    root_name = root_node[\"name\"]\n",
    "    children_predicates = parent_to_child_map.get(root_name, [])\n",
    "    if children_predicates:\n",
    "        children = []\n",
    "        for child_predicate in children_predicates:\n",
    "            child_node = {\"name\": child_predicate, \"parent\": root_name}\n",
    "            child_node = get_tree_node_recursive(child_node, parent_to_child_map)\n",
    "            children.append(child_node)\n",
    "        root_node[\"children\"] = sorted(children, key=lambda x: x[\"name\"])\n",
    "    return root_node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "6236a44c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2d/1xhwsfs15qb37bv4m62xn0tc0000gp/T/ipykernel_67537/2714310493.py:5: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  dataframe = dataframe.append(record, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "ptd = load_predicate_tree_data(\"master\")\n",
    "dataframe = pandas.DataFrame()\n",
    "for d in ptd[0]:\n",
    "    record = pandas.json_normalize(d)\n",
    "    dataframe = dataframe.append(record, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "74b35891",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_predicate_to_trapi_format(english_predicate: str) -> str:\n",
    "    # Converts a string like \"treated by\" to \"treated_by\"\n",
    "    return english_predicate.replace(' ', '_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "474f0b54",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
